cities/modeling/model_components.py

```diff
@@ -0,0 +1,251 @@
+
+    # a but convoluted, but groups might be missing and sometimes
+    # vars are allowed to be None
+    n_cat = None
+    if N_categorical > 0:
+        for value in categorical.values():
+            if value is not None:
+                n_cat = value.shape[0]
+                break
+
+    n_con = None
+    if N_continuous > 0:
+        for value in continuous.values():
+            if value is not None:
+                n_con = value.shape[0]
+                break
+
+    if N_categorical > 0 and N_continuous > 0:
+        if n_cat != n_con:
+            raise ValueError(
+                "The number of categorical and continuous data points must be the same"
+            )
+
+    n = n_cat if n_cat is not None else n_con
+
+    if n is None:
+        raise ValueError("Both categorical and continuous dictionaries are empty.")
+
+    return N_categorical, N_continuous, n
+
+
+def categorical_contribution(
+    categorical: Dict[str, torch.Tensor],
+    child_name: str,
+    leeway: float,
+    categorical_levels: Optional[Dict[str, torch.Tensor]] = None,
+) -> torch.Tensor:
+
+    categorical_names = list(categorical.keys())
+
+    if categorical_levels is None:
+        categorical_levels = {
+            name: torch.unique(categorical[name]) for name in categorical_names
+        }
+
+    weights_categorical_outcome = {}
+    objects_cat_weighted = {}
+
+    for name in categorical_names:
+        weights_categorical_outcome[name] = pyro.sample(
+            f"weights_categorical_{name}_{child_name}",
+            dist.Normal(0.0, leeway).expand(categorical_levels[name].shape).to_event(1),
+        )
+
+        weights_batch_dims = weights_categorical_outcome[name].shape[:-1]
+        # n = categorical[name].shape[-1]
+
+        # FIXME
+        #  Problem: if the categorical variable is being conditioned on per unit (e.g. in an ITE), then
+        #   it won't be plated according to the sample plate, which means that the gather below has to
+        #   broadcast those levels across the plated weights_categorical_outcome samples.
+        # HACKy solution:
+        # We can tell when this happens if weights_categorical_outcome batch shape doesn't match
+        #  the categorical shape. In that case, we have to manually tile to broadcast the gather.
+        # Otherwise, we use the view method.
+        if categorical[name].ndim == 1:  # we are conditioning.
+            weight_indices = torch.tile(
+                categorical[name].view(*((1,) * len(weights_batch_dims)), -1),
+                dims=(*weights_batch_dims, 1)
+            )
+        else:
+            weight_indices = categorical[name].view(*weights_batch_dims, -1)
+
+        objects_cat_weighted[name] = torch.gather(
+            weights_categorical_outcome[name],
+            dim=-1,
+            index=weight_indices
+        )
+
+    values = list(objects_cat_weighted.values())
+    for i in range(1, len(values)):
+        values[i] = values[i].view(values[0].shape)
```

Comment:
What's this doing? Seems extraneous cz all the `objects_cat_weighted` values should have the same shape already?